Enable single file analyzer in the runtime

tlakollo opened the pull request:
Delete ruleset for single file related warnings IL3000 and IL3001
Add ILLinkRoslyn analyzer and EnableSingleFileAnalyzer flag
Resolve warnings on single file dangerous patterns by:
- Annotating with RequiresAssemblyFiles
- Suppressing the warning
- Fixing the issue on code
- Opening an issue to track fix and either annotate/suppress the current behavior

Commit:
Enable single file analyzer in the runtime
Resolve warnings on single file dangerous patterns by:
- Annotating with RequiresAssemblyFiles
- Suppressing the warning
- Fixing the issue on code
- Opening an issue to track fix and either annotate/suppress the current
  behavior

tlakollo committed:
Merge branch 'main' into SingleFileAnalyzerOnRuntime

tlakollo commented:
Sending to CI to catch possible misses caused by:
- Code for an OS different from Windows
- Mono builds
Also, APICompat didn't fail on my computer although I added several Custom Attributes so I'm letting the CI run directly on the changes

MichalStrehovsky commented on src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextLoader.cs:
```diff
@@ -103,6 +105,7 @@ private DependencyContext LoadContext(IDependencyContextReader reader, string lo
             return null;
         }
 
+        [RequiresAssemblyFiles (Message = "Calls GetDepsJsonPath")]
```
Also elsewhere. This is not the formatting we use in .NET. This is the formatting that illink repo uses just to upset people.

```suggestion
        [RequiresAssemblyFiles(Message = "Calls GetDepsJsonPath")]
```

vitek-karas commented on src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContext.cs:
```diff
@@ -74,6 +75,7 @@ public DependencyContext Merge(DependencyContext other)
                 );
         }
 
+        [RequiresAssemblyFiles(Message = "Calls GetDepsJsonPath")]
```
This should be propagated to the public API (Default property) - I would expect the analyzer to generate some warnings around this. If not it's a bug in the analyzer which we should fix.

tlakollo replied:
I tried a little bit today to debug this, seems like a more complicated scenario than I expected. The analyzer is only able to see System.Lazy<DependencyContext>.Value.get but I have not found a way to connect it with the function is calling (LoadDefault). Also, I tested a similar scenario to see how does the linker behaves using RequiresUnreferencedCode and I found that they both have the same testing hole and don't produce the warning.
```C#
public static void Main ()
{
  TestCallsLazyInitializer ();
}

public static Lazy<RequiresUnreferencedCodeCapability> lazyRUC = new Lazy<RequiresUnreferencedCodeCapability> (InitRUC);

[RequiresUnreferencedCode ("Message for --RequiresUnreferencedCodeWithLazyInit--")]
public static RequiresUnreferencedCodeCapability InitRUC ()
{
	RequiresUnreferencedCodeCapability objectRUC = new RequiresUnreferencedCodeCapability ();
	return objectRUC;
}

[ExpectedWarning ("IL2026", "--RequiresUnreferencedCodeWithLazyInit--")]
public static void TestCallsLazyInitializer ()
{
	RequiresUnreferencedCodeCapability objectRUC = lazyRUC.Value;
}
```

MichalStrehovsky replied:
That looks like https://github.com/mono/linker/issues/1912 and related.

We should get a warning in the static constructor, when the Lazy instance is getting constructed.

tlakollo replied:
With https://github.com/mono/linker/pull/1964 now the Lazy call showed a warning, but since it warns on an implicit static constructor (I cannot annotate the static constructor unless I implement it) and then gets assigned to a field (which is a target that RequiresAssembly files don't support) I couldn't mark it with RequiresAssemblyFiles. I ended up refactoring the code and not make use of the Lazy call.

Commit:
Add IL3002 to CodeAnalysis.test.ruleset to avoid warning on tests
Add additional wasm pattern

Commit:
Merge branch 'SingleFileAnalyzerOnRuntime' of https://github.com/tlakollo/runtime into SingleFileAnalyzerOnRuntime

Commit:
Fix formating of CustomAttribute

Commit:
Fix introduced typo

MichalStrehovsky commented on src/libraries/Microsoft.NETCore.Platforms/src/AssemblyResolver.cs:
```diff
@@ -38,7 +38,10 @@ private static Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEven
 
             // look next to requesting assembly
+            // The code has a fallback to use AppDomain.CurrentDomain.BaseDirectory so is not single file dangerous
+#pragma warning disable IL3000 // Avoid accessing Assembly file path when publishing as a single file
```
Would it be possible to use the `UnconditionalSuppressMessage` attribute to do these suppressions?

From my past experience, every time someone hit an issue with APIs like `Assembly.Location` and couldn't figure it out, it was in a third party NuGet. If they hit it in their own code, they could usually just debug into it. E.g. the CoreRT repo is full of them; here's one: https://github.com/dotnet/corert/issues/7756.

I think there's a lot of value in eventually detecting this in compiled assemblies as well (either with illink or some other tool). We're not doing that right now. Those tools will not see the `#pragama warning` suppressions, so we'll have to redo it.

E.g. I could add this detection to the NativeAOT compiler now, but if the suppressions in the framework are only for the analyzer, there's little point to do that.

tlakollo replied:
Changed to UnconditionalSuppressedMessage since a new version of arcade that include the roslyn changes was merged to runtime

Commit:
Merge branch 'main' of https://github.com/dotnet/runtime into SingleFileAnalyzerOnRuntime

Commit:
Merge branch 'main' of https://github.com/dotnet/runtime into SingleFileAnalyzerOnRuntime

Commit:
Merge branch 'main' of https://github.com/dotnet/runtime into SingleFileAnalyzerOnRuntime
Change pragmas for UnconditionalSuppressMessage
Suppress wasm error and open an issue about it

Commit:
Fix some formatting/comments/extra pragma

Commit:
Fix UnconditionalSuppressMessage when TFM != NetCoreAppCurrent

Commit:
Add more UnconditionalSuppressMessage support for TFM !=
NetCoreAppCurrent

Commit:
Fix incompatibility in EventLog.csproj

Commit:
Add System.Diagnostics.Tools if TFM is netcoreapp3.0 on
System.Reflection.MetadataLoadContext

vitek-karas commented on src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextLoader.cs:
```diff
@@ -50,6 +51,7 @@ private static Stream GetResourceStream(Assembly assembly, string name)
             return assembly.GetManifestResourceStream(name);
         }
 
+        [RequiresAssemblyFiles(Message = "Calls GetDepsJsonPath")]
```
Public APIs should have some meaningful and actionable message for the attribute.
Granted there's no action in this case (we simply don't support single-file with this API), but at least it should talk in terms of the user's experience and not implementation details.

vitek-karas commented on src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextLoader.cs:
```diff
@@ -125,6 +128,7 @@ private DependencyContext LoadAssemblyContext(Assembly assembly, IDependencyCont
             return null;
         }
 
+        [RequiresAssemblyFiles(Message = "Needs additional support for reading .deps.json in single file", Url = "https://github.com/dotnet/runtime/issues/41265")]
```
I would not use the `Url` property for pointing to active issues...
Typically what we do in other places we add a const string which says that the feature is not supported with single-file and what workaround there is (if any) and then use that string in all the attributes which were added because of that feature.

Implementation details can be added in comments if needed.

vitek-karas commented on src/libraries/Microsoft.NETCore.Platforms/src/AssemblyResolver.cs:
```diff
@@ -27,6 +28,8 @@ public static void Enable()
         }
 
+        [UnconditionalSuppressMessage("Single file", "IL3000:Avoid accessing Assembly file path when publishing as a single file",
+            Justification = "The code has a fallback to use AppDomain.CurrentDomain.BaseDirectory so is not single file dangerous")]
```
```suggestion
            Justification = "The code has a fallback to use AppDomain.CurrentDomain.BaseDirectory so it will work correctly in single-file")]
```

vitek-karas commented on src/libraries/Microsoft.XmlSerializer.Generator/src/Sgen.cs:
```diff
@@ -206,6 +207,8 @@ private int Run(string[] args)
         }
 
+        [UnconditionalSuppressMessage("Single file", "IL3000: Avoid accessing Assembly file path when publishing as a single file",
+            Justification = "Code has a fallback")]
```
```suggestion
            Justification = "Code has a fallback and thus works in single-file")]
```

vitek-karas commented on src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/AssemblyCatalog.cs:
```diff
@@ -543,6 +543,8 @@ private void ThrowIfDisposed()
             $"{GetType().Name} (Assembly=\"{Assembly.FullName}\")";   // NOLOC
 
+        [UnconditionalSuppressMessage("Single file", "IL3000: Avoid accessing Assembly file path when publishing as a single file",
+            Justification = "Setting a CodeBase is not single file dangerous")]
```
In general I would avoid "dangerous"... we can use for example "single file compatible" or similar terminology.

vitek-karas commented on src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContext.cs:
```diff
@@ -57,6 +58,8 @@ private string GetLocalPath(string fileName)
             return uri.LocalPath + uri.Fragment;
         }
 
+        [UnconditionalSuppressMessage("Single file", "IL3000: Avoid accessing Assembly file path when publishing as a single file",
+            Justification = "Suppressing the warning until gets fixed, see https://github.com/dotnet/runtime/issues/50821")]
```
What would happen if we didn't suppress this - how much would it spread (propagating the attribute to affected public APIs).
The functionality will not work in single-file - so silencing it until it gets fixed is wrong. The warnings are there for this specific reason.

tlakollo replied:
Adding [RequiresAssemblyFiles] in this function doesn't propagate to other places. I think is a limitation in the analyzer, because the attribute is on an override, but the methods call a virtual function in [LicenseContext.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseContext.cs)
If propagated to the callers, then [GetCurrentContextInfo](https://github.com/dotnet/runtime/blob/main/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.LicenseInteropHelper.cs#L110) uses LicenseUsageMode.Designtime and it's called by COM so it doesn't propagate more.
The other caller is [GetLicense](https://github.com/dotnet/runtime/blob/main/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicFileLicenseProvider.cs#L43) which uses LicenseUsageMode.Runtime, for me is uncertain if that will result in calling the override in DesigntimeLicenseContext. But that method has about 10 different call sites that would need to be annotated most of them in System.ComponentModel.LicenseManger.

vitek-karas replied:
> and it's called by COM so it doesn't propagate more.

Well - it doesn't propagate in the analyzer - but functionally it does propagate - so this is the killer argument for me. It would basically mean we should potentially annotate the entire COM this way.

Regarding the analyzer: We should add a similar check to what RequiresUnreferencedCode has (at least in the linker) - that is, all methods in a virtual override chain should have the same annotations. So if an override has the attribute, the base method should have it as well, and vice-versa. If this is not the case, please file a bug for this.

tlakollo replied:
I opened https://github.com/mono/linker/issues/1985 to track work for virtual calls in the analyzers

vitek-karas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -115,6 +115,7 @@ public virtual IEnumerable<Type> ExportedTypes
         public virtual object[] GetCustomAttributes(bool inherit) { throw NotImplemented.ByDesign; }
         public virtual object[] GetCustomAttributes(Type attributeType, bool inherit) { throw NotImplemented.ByDesign; }
 
+        [RequiresAssemblyFiles(Message = "Calls 'AssemblyName.EscapeCodeBase(string)' which is a dangerous single file pattern")]
```
There's no need to repeat the method's name - the warning generated by the analyzer should already contain that. At least that's how RequiresUnreferencedCode works, I would assume RequiresAssemblyFiles would work the same way.

Avoid "dangerous". There should be a suggestion to how to fix the problem.

vitek-karas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -152,6 +153,7 @@ public virtual IEnumerable<Type> ExportedTypes
         public virtual Assembly GetSatelliteAssembly(CultureInfo culture, Version? version) { throw NotImplemented.ByDesign; }
 
         public virtual FileStream? GetFile(string name) { throw NotImplemented.ByDesign; }
+        [RequiresAssemblyFiles(Message = "'System.Reflection.Assembly.GetFiles(bool)' will throw for assemblies embedded in a single-file app", Url = "https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/il3001")]
```
URL pointing to the warning should not be in the `Url` property - VS will do that automatically and everywhere else people can easily search for the warning.
The `Url` property is meant to point to docs specifically about this method - if there's any.

vitek-karas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs:
```diff
@@ -752,6 +756,8 @@ private static void OnAssemblyLoad(RuntimeAssembly assembly)
 
         [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
             Justification = "Satellite assemblies have no code in them and loading is not a problem")]
+        [UnconditionalSuppressMessage("Single file", "IL3000: Avoid accessing Assembly file path when publishing as a single file",
+            Justification = "This call is fine because native call runs before this and checks BindSatelliteResourceFromBundle")]
```
So the Justification is that this should never get called for single-file bundled assemblies, right?

vitek-karas replied:
In that case the method should still handle null from `Assembly.Location` in some way - at the very least an assert, but maybe just return null if it's null. Currently it will NullRef which is pretty bad.

tlakollo replied:
Returning null will be just hiding a problem, I will put an assert instead

tlakollo replied:
Actually the code is already handling null
```C#
string? parentDirectory = Path.GetDirectoryName(parentAssembly.Location);
if (parentDirectory == null)
    return null;
```
GetDirectoryName checks for null and string.Empty and in that case returns null

vitek-karas commented:
@eerhardt What is your take on some of the cases where we suppress warnings and file issues on them. I'm not a fan since those are real problems we're hiding. On the other hand propagating these warnings feels like it would affect too large of an API surface.
For trimming we took the approach of warnings anyway I think, so I don't want to set a bad precedent if we want to stick to that behavior.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextLoader.cs:
```diff
@@ -125,6 +128,7 @@ private DependencyContext LoadAssemblyContext(Assembly assembly, IDependencyCont
             return null;
         }
 
+        [RequiresAssemblyFiles(Message = "Needs additional support for reading .deps.json in single file", Url = "https://github.com/dotnet/runtime/issues/41265")]
```
This should be reverted. If you look at the code of this method, it is 100% safe and doesn't "RequiresAssemblyFiles".

```C#
            // Assemblies loaded in memory (e.g. single file) return empty string from Location.
            // In these cases, don't try probing next to the assembly.
            string assemblyLocation = assembly.Location;
            if (string.IsNullOrEmpty(assemblyLocation))
            {
                return null;
            }
```

tlakollo replied:
After https://github.com/dotnet/runtime/pull/41430 was merged we don't crash but we don't have a way to actually read the content of the application's .deps.json file which is currently not accessible in any way. So I think this pattern is still single file incompatible, I guess @vitek-karas  has more insight on this

vitek-karas replied:
I think we should keep the warning here - it's true that there will be callers which handle null return value and those would just suppress the warning, but most callers I've seen don't expect null, so in this specific case I think it makes sense to keep the warning. Also because it basically says - calling this for single-file is a no-op, the feature just doesn't work at all in single-file.

eerhardt commented:
> @eerhardt What is your take on some of the cases where we suppress warnings and file issues on them. I'm not a fan since those are real problems we're hiding. On the other hand propagating these warnings feels like it would affect too large of an API surface.
For trimming we took the approach of warnings anyway I think, so I don't want to set a bad precedent if we want to stick to that behavior.

I think there needs to be a better pattern here. The analyzer should be able to tell if the code is checking `assembly.Location` for empty string, and if the code does, the analyzer shouldn't issue a warning.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyModel/src/Microsoft.Extensions.DependencyModel.csproj:
```diff
@@ -30,4 +30,8 @@
     <ProjectReference Include="$(LibrariesProjectRoot)System.Text.Json\src\System.Text.Json.csproj" />
   </ItemGroup>
 
+  <ItemGroup Condition="'$(TargetFramework)' != '$(NetCoreAppCurrent)'">
```
Even though this will be reverted, I don't think it is appropriate to check for TargetFrameworks that a .csproj doesn't build for. This project builds for `netstandard2.0;net461`. Thus this condition is always `true`.

eerhardt commented on src/libraries/Microsoft.XmlSerializer.Generator/src/Microsoft.XmlSerializer.Generator.csproj:
```diff
@@ -7,8 +7,15 @@
     <Compile Include="Sgen.cs" />
   </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' != '$(NetCoreAppCurrent)'">
```
This project only builds for `netstandard2.0`. This condition should be removed.

Repeated elsewhere.

eerhardt commented on src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/AssemblyCatalog.cs:
```diff
@@ -543,6 +543,8 @@ private void ThrowIfDisposed()
             $"{GetType().Name} (Assembly=\"{Assembly.FullName}\")";   // NOLOC
 
+        [UnconditionalSuppressMessage("Single file", "IL3000: Avoid accessing Assembly file path when publishing as a single file",
+            Justification = "Setting a CodeBase is not single file dangerous")]
```
If setting a CodeBase isn't a problem - then why is it annotated as such?

tlakollo replied:
The analyzer treats the whole property as dangerous because getting information from it will be empty in single-file

vitek-karas replied:
RequiresUnreferencedCode can be applied to just one of the accessors - doesn't it work the same for RequiresAssemblyFiles as well?

tlakollo replied:
The analyzer recognizes a special range of dangerous patterns in which RequiresAssemblyFiles is irrelevant since they will generate IL3000 and not IL3002 (even if annotated with RequiresAssemblyFiles). I'm looking into how to make it warn only on the getter.

eerhardt replied:
RequiresUnreferencedCode can't be applied to a property. It is only valid on a method. So you need to annotate 1 or both accessors if you want it on a property.

tlakollo replied:
Actually, I got rid of CodeBase/EscapedCodeBase in the analyzer since they are [obsolete patterns](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/ref/System.Runtime.cs)
https://github.com/mono/linker/pull/1966

tlakollo replied:
There was a discussion in mono/linker#1966 the plan is:
Suppress the message and annotate CodeBase getter with RequiresAssemblyFiles
After the annotation flows, we delete the code in the analyzer that special cases CodeBase and EscapedCodeBase along with tests
After deletion of the code in the analyzer flows to runtime, we can delete the suppression in the runtime

eerhardt commented on src/tasks/WasmAppBuilder/WasmLoadAssembliesAndReferences.cs:
```diff
@@ -29,6 +29,7 @@ public class WasmLoadAssembliesAndReferences : Task
 
     private SortedDictionary<string, Assembly> _assemblies = new();
 
+    [RequiresAssemblyFiles(Message = "Using 'System.Reflection.Assembly.Location' will make ReferencedAssemblies to be empty")]
```
This is a bit odd - this is an MSBuild task. What good is annotating it with `RequiresAssemblyFiles`? No upstream caller (i.e. MSBuild) is going to care about this attribute.

I assume we can just suppress the warning here - or maybe even turn off the `EnableSingleFileAnalyzer`, since this Task will only ever run in the context of MSBuild, and not a single file app.

vitek-karas replied:
If the project will never run in single-file we should disable the analyzer for that project. Adding annotations for something which can never happen is just noise.

tlakollo commented:
> I think there needs to be a better pattern here. The analyzer should be able to tell if the code is checking `assembly.Location` for empty string, and if the code does, the analyzer shouldn't issue a warning.

@eerhardt I'm still confused by this comment, the EnableSingleFileAnalyzer does not publish the app, it's just recognizing patterns in source code. I don't think there is a way for the analyzer to know if assembly.Location will return an empty string or not since nothing is bundled.

Commit:
Address some PR comments

vitek-karas commented:
> I don't think there is a way for the analyzer to know if assembly.Location will return an empty string or not since nothing is bundled.

Roslyn can currently recognize for example this:
```C#
string? filepath = typeof(object).Assembly.Location;
if (filepath == null) return;
filepath.ToString();  // This will not warn since Roslyn recognizes the if above and knows that even if it's nullable string, the value can never be null here.
```

The idea is similar - the single-file analyzer could recognize that the code handles null/empty-string by checking for them before using the value for anything else. If the code has a check then it already knows that the value can be null/empty-string and that's all we want to achieve with the warning.

vitek-karas commented on src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContext.cs:
```diff
@@ -75,7 +75,7 @@ public DependencyContext Merge(DependencyContext other)
         }
 
-        [RequiresAssemblyFiles(Message = "Calls GetDepsJsonPath")]
+        [RequiresAssemblyFiles(Message = "The use of DependencyContextLoader is not supported when publishing as single-file")]
```
These should be user facing messages. While this is technically correct it doesn't help much to the user.

```C#
// IL3002: The use of DependencyContextLoader is not supported when publishing as single-file.
DependencyContext.Default.ToString();
```

User: What??? I'm not calling `DependencyContextLoader`, I'm calling `DependencyContext.Default`... this must be a bug - I'll ignore it.

The message should read something like:
"DependencyContext for an assembly from a application published as single-file is not supported. The method will return `null`. Make sure the calling code can handle this case."

MichalStrehovsky commented:
> I think there needs to be a better pattern here. The analyzer should be able to tell if the code is checking assembly.Location for empty string, and if the code does, the analyzer shouldn't issue a warning.

The analyzer warns if there's potential for a difference in behavior at runtime after compiling into a single file. I don't think checking for null should suppress the warning.

If what the code will do is indeed identical, then why not just delete the path that checks `.Location` and use the `null` codepath always? The `null` codepath might be specific to assemblies loaded from byte arrays/Ref.Emitted, and it still might be unexpected for the null codepath to be taken in single file - the behavior needs to be reviewed and suppressed as needed.

eerhardt commented:
> The analyzer warns if there's potential for a difference in behavior at runtime after compiling into a single file. I don't think checking for null should suppress the warning.
> If what the code will do is indeed identical, then why not just delete the path that checks .Location and use the null codepath always? The null codepath might be specific to assemblies loaded from byte arrays/Ref.Emitted, and it still might be unexpected for the null codepath to be taken in single file - the behavior needs to be reviewed and suppressed as needed.

The issue is there is no way for a developer to write their code without adding a suppression. The only option is to suppress. There is no "capability API" for this.

agocke commented:
> The issue is there is no way for a developer to write their code without adding a suppression. The only option is to suppress. There is no "capability API" for this.

So, we talked about this in design review a while back. https://youtu.be/TPCjoYyp6e0?list=PL1rZQsJPBU2S49OQPjupSJF-qeIEz9_ju&t=953

The problem was: there are some situations in which the "capability" you're checking for is just whether or not a particular assembly is on disk, like Assembly.Location. On the other hand, you may also need to know if anything is in a single-file bundle, not just a particular file.

I think I have the same feeling I did when we did the design review, namely that I'm not sure what the right API is here but I'm open to proposals.

If we wanted to specifically make an exception that `assembly.Location == string.Empty` doesn't warn -- that seems fine. No flow analysis necessarily, just that single expression seems fine. We should also probably special case `assembly.Location.Length != 0`.

agocke commented:
Also, this could all be solved by adding an API, `assembly.IsFromMemory` property that avoids all special-casing and doesn't warn.

Commit:
Merge branch 'main' of https://github.com/dotnet/runtime into SingleFileAnalyzerOnRuntime

Commit:
Change code in DependencyContext to not use Lazy
PR comments

vitek-karas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -271,7 +271,8 @@ public static Assembly LoadFile(string path)
         [RequiresUnreferencedCode("Types and members the loaded assembly depends on might be removed")]
-        [RequiresAssemblyFiles(Message = "Calling 'System.Reflection.Assembly.LoadFromResolveHandler' will return a null assembly for assemblies embedded in a single-file app")]
+        [UnconditionalSuppressMessage("SingleFile", "IL3000:Avoid accessing Assembly file path when publishing as a single file",
+            Justification = "requestingAssembly contains a path to the assembly location outside of the single file application")]
```
This is the case where we should never get here for an assembly from single-file bundle, right? I would reword it slightly to make that more obvious.

tlakollo replied:
Yeah, since it gets called by LoadFrom(string) we load from a string with a location outside the bundle, will change the wording

vitek-karas commented on src/libraries/System.Reflection.MetadataLoadContext/src/System.Reflection.MetadataLoadContext.csproj:
```diff
@@ -1,4 +1,4 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
```
Revert this file

vitek-karas commented on src/libraries/Microsoft.XmlSerializer.Generator/src/Sgen.cs:
```diff
@@ -293,6 +296,9 @@ private void GenerateFile(List<string> typeNames, string assemblyName, bool prox
                 }
             }
+            string nameOrLocation = assembly.Location;
+            if (nameOrLocation == string.Empty)
```
Should this also check for `null`?

tlakollo replied:
The only implementation I found in the runtime was in [RuntimeAssembly.cs](https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs#L457) and suppresses the warning for nullable. Although I'm not very sure if the [Qcall ](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/assemblynative.cpp#L410) will not return null

tlakollo replied:
I change it either way to string.IsNullOrEmpty pattern

agocke replied:
`Assembly.Location` says it returns non-null string. Unless there's any particular reason to think there's a bug, I think we can trust that this never returns null.

vitek-karas commented on src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/AssemblyCatalog.cs:
```diff
@@ -543,6 +543,8 @@ private void ThrowIfDisposed()
         private string GetDisplayName() =>
             $"{GetType().Name} (Assembly=\"{Assembly.FullName}\")";   // NOLOC
 
+        [UnconditionalSuppressMessage("Single file", "IL3000: Avoid accessing Assembly file path when publishing as a single file",
+            Justification = "Setting a CodeBase is single file compatible")]
```
Did you look at only making the getter for CodeBase "unsafe" and leave the setter without an attribute?

tlakollo replied:
Yes, the annotation is in the getter of codebase in the file `src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs` in this PR.
I didn't change the analyzer behavior for now since it will mean more work to special case the codebase getter in the analyzer. The plan is to annotate first, then get rid of the special casing in the analyzers for CodeBase/EscapedCodeBase since they are deprecated and provide a misleading warning, and once the analyzer doesn't check for those we can delete the UnconditionalSuppressMessage and tests in the linker. Also cannot remove it for now since the annotations are not in place and we wouldn't warn.

vitek-karas reviewed (approved):
Looks good

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyModel/src/Microsoft.Extensions.DependencyModel.csproj:
```diff
@@ -25,4 +25,9 @@
   </ItemGroup>
+
+  <ItemGroup>
+    <Compile Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\RequiresAssemblyFilesAttribute.cs" />
```
Can this go up with the `HashHelpers` file above?

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContext.cs:
```diff
@@ -3,14 +3,17 @@
 {
     public class DependencyContext
     {
-        private static readonly Lazy<DependencyContext> _defaultContext = new Lazy<DependencyContext>(LoadDefault);
```
Can you undo this change? The proposed change doesn't have the same semantics as the existing code - you can get 2 DependencyContext objects created.

tlakollo replied:
Ahhh yeah thread safety, I will suppress the warning on the field and add RequiresAssemblyFiles on the Default function

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContext.cs:
```diff
@@ -46,7 +49,19 @@ public class DependencyContext
+        {
+            get
+            {
+                if (_isDefaultContextInitialized == false) {
```
opening braces go on new lines in this repo.

eerhardt commented on src/libraries/Microsoft.XmlSerializer.Generator/src/Microsoft.XmlSerializer.Generator.csproj:
```diff
@@ -7,8 +7,15 @@
   </ItemGroup>
+
+  <ItemGroup>
+    <Compile Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\UnconditionalSuppressMessageAttribute.cs" />
```
Can you just collapse this up into the previous ItemGroup?

eerhardt commented on src/libraries/System.Configuration.ConfigurationManager/src/System.Configuration.ConfigurationManager.csproj:
```diff
@@ -258,4 +258,7 @@
     <Reference Include="System.Configuration" />
   </ItemGroup>
+  <ItemGroup>
+    <Compile Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\UnconditionalSuppressMessageAttribute.cs" />
```
Can this go into the above ItemGroup with the rest of the Compiles?

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs:
```diff
@@ -260,6 +263,7 @@ public static bool ReferenceMatchesDefinition(AssemblyName? reference, AssemblyN
             return refName.Equals(defName, StringComparison.OrdinalIgnoreCase);
         }
 
+        [RequiresAssemblyFiles(Message = "The code will throw for assemblies embedded in a single-file app")]
```
It will? where?

tlakollo replied:
Yikes, I thought the documentation said something about it, but it was [CodeBase ](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.codebase?view=net-5.0) the one that throws. I will change the message to reflect string.Empty instead.

tlakollo replied:
I created a program to verify the behavior
```
var a = Assembly.GetExecutingAssembly().GetName();
_ = a.CodeBase;  // returns string.Empty
_ = a.EscapedCodeBase;  // returns string.Empty
var b = Assembly.GetExecutingAssembly();
_ = b.CodeBase;  // throws and generates obsolete warning
_ = b.EscapedCodeBase; // throws and generates obsolete warning
```
It depends on if you call CodeBase/EscapedCodeBase from AssemblyName.cs or Assembly.cs the behavior it will take. 
Calling EscapedCodeBase from assembly will generate the following stack trace
```
Unhandled exception. System.NotSupportedException: CodeBase is not supported on assemblies loaded from a single-file bundle.
   at System.Reflection.RuntimeAssembly.get_CodeBase()
   at System.Reflection.Assembly.get_EscapedCodeBase()
   at EscapedCodeBase.Program.Main(String[] args)
```
Here we can see that uses the [RuntimeAssembly ](https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs#L94) implementation which throws.
In short, I need to modify the messages to reflect the behavior

eerhardt commented on src/libraries/System.Reflection.Context/src/System.Reflection.Context.csproj:
```diff
@@ -64,4 +64,7 @@
-</Project>
\ No newline at end of file
+  <ItemGroup>
+    <Compile Include="$(CoreLibSharedDir)System\Diagnostics\CodeAnalysis\RequiresAssemblyFilesAttribute.cs" />
```
Can this go in the above ItemGroup?

Commit:
PR feedback
Change Single file for category in the analyzer SingleFile

Commit:
Redo changes in MetadataLoadContext.csproj after checking out the main
file

Commit:
Merge branch 'main' of https://github.com/dotnet/runtime into SingleFileAnalyzerOnRuntime

Commit:
Delete extra code

MichalStrehovsky commented on src/libraries/Microsoft.XmlSerializer.Generator/src/Microsoft.XmlSerializer.Generator.csproj:
```diff
@@ -7,8 +7,12 @@
     <OutputType>Exe</OutputType>
     <TargetFrameworks>netstandard2.0</TargetFrameworks>
   </PropertyGroup>
+
```
This is a command line tool (`<OutputType>Exe</OutputType>`) - do we expect SGEN to ever be compiled as single file? I think we should just disable the analyzer for this instead of making changes to make it single file friendly

tlakollo replied:
Disabled SingleFileAnalyzer and reverted the changes

Commit:
Disable SingleFileAnalyzer for Microsoft.XmlSerializer.Generator and
revert changes

tlakollo committed:
Merge branch 'main' into SingleFileAnalyzerOnRuntime

tlakollo commented:
Since this PR has been passing a couple of times the only question that I have is why API compat doesn't warn on these new attributes, I didn't add any code to the ref assemblies shouldn't this suppose to break the build?
cc @safern

Commit:
Missing message change

Commit:
Merge branch 'SingleFileAnalyzerOnRuntime' of https://github.com/tlakollo/runtime into SingleFileAnalyzerOnRuntime

safern commented on src/libraries/System.Diagnostics.EventLog/src/System.Diagnostics.EventLog.csproj:
```diff
@@ -128,6 +128,9 @@
     <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Win32.Registry\src\Microsoft.Win32.Registry.csproj" />
     <ProjectReference Include="$(LibrariesProjectRoot)System.Threading.AccessControl\src\System.Threading.AccessControl.csproj" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetFramework)' != '$(NetCoreAppCurrent)-windows'">
```
We usually don't use the `-os` portion for these conditions.

tlakollo replied:
Changed to !$(TargetFramework.StartsWith('$(NetCoreAppCurrent)'))

safern commented:
> Since this PR has been passing a couple of times the only question that I have is why API compat doesn't warn on these new attributes, I didn't add any code to the ref assemblies shouldn't this suppose to break the build?

Yes this should be failing. Let me try and reproduce this locally and debug it to see what's going on.

tlakollo committed:
PR feedback

safern commented:
It seems like the APICompat issue was a regression I introduced when adding support to diff attributes on type parameters and method parameters. I put up a fix for this regression: https://github.com/dotnet/arcade/pull/7376

tlakollo committed:
Merge branch 'main' of https://github.com/dotnet/runtime into SingleFileAnalyzerOnRuntime

tlakollo committed:
Merge branch 'main' of https://github.com/dotnet/runtime into SingleFileAnalyzerOnRuntime

safern commented:
@tlakollo the version of API Compat containing my fix is being updated in: https://github.com/dotnet/runtime/pull/52440. So once that is merged you should be unblocked to do a build rebasing on top of main and getting errors of attribute miss matches.

tlakollo committed:
Merge branch 'main' of https://github.com/dotnet/runtime into SingleFileAnalyzerOnRuntime

tlakollo committed:
Change UnconditionalSuppressMessage for SuppressMessage since SingleFile
analysis is not made by a GlobalAnalysis tool and there is no reason to
have extra IL with the attribute
Add ref assembly attributes
Fix a single file incompatible pattern issue found in System.IO.IsolatedStorage

tlakollo commented:
I changed the attributes from UnconditionalSuppressMessage to SuppressMessage since there is no reason at this moment to have the extra IL, since the analysis is being conducted via analyzers is not necessary.

tlakollo commented:
For the System.IO.IsolatedStorage I will need review and guidance on how to test that my changes don't change the way the app behaves for both single-file and non single-file 
Adding @JeremyKuhne @adamsitnik @carlossanlop @Jozkee @jeffhandley

MichalStrehovsky commented:
> I changed the attributes from UnconditionalSuppressMessage to SuppressMessage since there is no reason at this moment to have the extra IL, since the analysis is being conducted via analyzers is not necessary.

This means NativeAOT will not be able to implement the single file warnings since the attributes need to be in IL as per https://github.com/dotnet/runtime/pull/50894#discussion_r610404380. SuppressMessage is no different from `#pragma warning disable`.

vitek-karas commented:
> This means NativeAOT will not be able to implement the single file warnings

Can we rely on the analyzers instead?
So far the single-file analysis can be done locally (there's nothing which requires global view), which makes it much more suitable for analyzers. The disadvantages are:
* It's "viral", in that any single component which is not single-file compatible can break the app, and so only global analyzers can tell if the app as a whole is compatible
* Analyzers are C#/VB only

For now we've been thinking about this as very similar to all the other analyzers - most specifically the platform analyzer, which is basically identical ("viral" as well). And the outcome was that unless we can solve this for all of the analyzers, it makes little sense to try to solve it for one specific one alone.

(Note: Trimming is different in that it needs to run on IL and needs to be global)

I'm a bit split on this one - it would help in NativeAOT with a more predictable experience... but to be fully predictable we would also have to support global analysis for things like platform compatibility and so on...

MichalStrehovsky commented:
I look at the single file analyzer as a stopgap until we have a better spot for this. I'll just quote from https://github.com/dotnet/runtime/pull/50894#discussion_r610404380.

"From my past experience, every time someone hit an issue with APIs like Assembly.Location and couldn't figure it out, it was in a third party NuGet. If they hit it in their own code, they could usually just debug into it. E.g. the CoreRT repo is full of them; here's one: dotnet/corert#7756."

Basically you get a lot more payoff from the analyzer warning if it warns you about code that is not yours. The null from Assembly.Location results in so many ArgumentNullException that I always start my investigations with that - and as someone fluent with ILSpy I can usually debug it in 2 minutes. Here's another https://github.com/grpc/grpc/issues/18188 loooong thread with a lot of angry/frustrated people. It would be much better if we could just issue a warning that the NuGet package they're using is not compatible with single file deployments than trying to get a message across a busy thread with a lot of frustrated people and package maintainers uninterested in fixing the issue of their users.

MichalStrehovsky commented:
(The interesting part of the GRPC thread starts at https://github.com/grpc/grpc/issues/18188#issuecomment-542712813 where what was a .NET Native bug that we fixed made forward progress to hit a GRPC bug)

vitek-karas commented:
OK - makes sense. I still hate the fact that we're solving these things as one-offs (platform analyzer has this problem as well, without any solution) - also this will only work in native AOT - JITed single-file will not show any warning and break anyway.
But I guess it makes sense to use `UnconditionalSuppressMessage` in the hopes that one day we will have a general global analyzer for these things (which would then use the attribute).

@tlakollo Sorry for the back and forth on this - can you please revert back to using `UnconditionalSuppressMessage` - and we should also stick with the code-fixer for it as well I guess. While there's very little use for it today (Native AOT is the only one), hopefully this will make it future proof.

tlakollo committed:
Merge branch 'main' of https://github.com/dotnet/runtime into SingleFileAnalyzerOnRuntime
Change SuppressMessage for UnconditionalSuppressMessage to support AOT
scenarios

tlakollo commented:
I was able to test that executing in `runtime\src\libraries\System.IO.IsolatedStorage\tests` the command `dotnet build -t:test -p:TestSingleFile=true` the end result with the previous code would be 235 failures out of 267 test with the same failure signature
235
```
[FAIL] System.IO.IsolatedStorage.RemoveTests.RemoveUserStoreForDomain
  System.NotSupportedException : CodeBase is not supported on assemblies loaded from a single-file bundle.
     at System.Reflection.RuntimeAssembly.get_CodeBase() in System.Private.CoreLib.dll:token 0x6004e08+0x1b
     at System.IO.IsolatedStorage.Helper.GetDefaultIdentityAndHash(Object& identity, String& hash, Char separator) in System.IO.IsolatedStorage.dll:token 0x60000af+0x24
     at System.IO.IsolatedStorage.IsolatedStorage.InitStore(IsolatedStorageScope scope, Type domainEvidenceType, Type assemblyEvidenceType) in System.IO.IsolatedStorage.dll:token 0x60000a6+0x1a
     at System.IO.IsolatedStorage.IsolatedStorageFile..ctor(IsolatedStorageScope scope) in System.IO.IsolatedStorage.dll:token 0x600002a+0x13
     at System.IO.IsolatedStorage.IsolatedStorageFile.GetStore(IsolatedStorageScope scope) in System.IO.IsolatedStorage.dll:token 0x6000050+0x1
     at System.IO.IsolatedStorage.IsolatedStorageFile.GetUserStoreForDomain() in System.IO.IsolatedStorage.dll:token 0x600004b+0x1
     at System.IO.IsolatedStorage.RemoveTests.RemoveUserStoreForDomain() in A:\Repos\SingleFileAnalyzerOnRuntime\runtime\src\libraries\System.IO.IsolatedStorage\tests\System\IO\IsolatedStorage\RemoveTests.cs:line 45
  Finished System.IO.IsolatedStorage.Tests, Version=6.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
```
After the code change the 267 test passed, although I'm not 100% sure that there is a test that exercises the particular scenario, I will update this PR with a change in runtime/src/libraries/tests.proj to enable running System.IO.IsolatedStorage tests as a single file.

tlakollo committed:
Merge branch 'main' of https://github.com/dotnet/runtime into SingleFileAnalyzerOnRuntime
Enable single file testing in System.IO.IsolatedStorage
